### DISCUSSION TITLE
Discussion on observation.py

### DISCUSSION THREAD

**Discussion by SEUCGX** [eleurent/highway-env#399] [Show and tell]
Thanks for providing a simple and excellent environment by the author.
I found some problems in using environment “highway-v0". I believe that in this scenario, it is better to choose Occupancy grid as the observation state than Kinematics. However, there are some problems.
(1) in line 312 of observation.py of current version, the information of "df" is actually the relative position, and the lane information of ego vehicle is ignored. This is fatal for vehicles on the leftmost and rightmost sides. They may make illegal actions because they do not know their lane information. I suggest adding lane information to the observation in any way to help ego-vehicle decision-making.

(2) line317-328 fill the vehicle information into the observation matrix in a circular manner. line327 does not check whether the current grid is occupied, which means that information may be overwritten. Due to the logic of the previous code, the cycle sequence in df follows the rule of "from near to far" according to ego-vehicle. In this case, if there are exactly two vehicles in one grid, the grid will be filled with vehicle information far away from ego-vehicles.  In some cases, this may lead to a vehicle being invisible, and may also cause a collision. On the contrary, it is reasonable for me to fill the information of this grid with vehicles closer to ego-vehicle. This will affect the results when making decisions. I suggest reversing the cycle order or adding a check.
When each grid is 20m (because the average speed of other vehicles is about 21m/s), the problem of two vehicles in one grid will not occur in most cases. But it happens if env.seed(33).

**Comment by eleurent** [comment]
Hi; thank you for the feedback!
(1) You're right. Maybe the lane can be somewhat inferred by looking at the distribution of other vehicles, but that's not really reliable. Maybe a simple solution is to add a channel showing the vehicles' lane index (must be indexed as an int). It would be better to have a lane occupancy grid though, just it requires a bit more work. 
(2) Good point!

**Reply by SEUCGX** [reply]
Thank you for your reply. I almost agree with you. However, considering that most of the values in the state space are float, it would be a little strange (but not necessarily infeasible) for lane information to be an int. The solution I've taken is to put the information about the ego-vehicle in its current position, i.e., obs[1:,0,1], except for the x-axis information,  which is really unimportant. Since I only used position and velocity information, my code is
normal_y=12
normal_vx_vy=[30 6]
obs[1,0,1]=self.observer_vehicle.destination[1]/normal_y
obs[2:,0,1]=(self.observer_vehicle.velocity/normal_vx_vy).

This is not the standard writing, but it easily solves the problem I encountered.
